File: scripts/test-task-logic.py
import os
import psycopg2
import json
import urllib.request
from dotenv import load_dotenv

load_dotenv()
DATABASE_URL = os.getenv("DATABASE_URL")
PYTHON_API_URL = "http://localhost:8000"

def get_connection():
    # Use require instead of verify-full to avoid local CA certificate issues
    url = DATABASE_URL.replace("sslmode=verify-full", "sslmode=require")
    return psycopg2.connect(url)

def post_predict(features):
    data = json.dumps(features).encode("utf-8")
    req = urllib.request.Request(f"{PYTHON_API_URL}/predict", data=data, headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req) as response:
        return json.loads(response.read().decode("utf-8"))

def test_flow():
    conn = get_connection()
    cur = conn.cursor()

    try:
        # 1. Create a fresh Deal
        print("--- Step 1: Create a fresh Deal ---")
        name = "Test Engagement Deal"
        amount = 50000
        stage = "Qualified to Buy"
        # Let's get a real orgId from the DB
        cur.execute("SELECT organization_id FROM users LIMIT 1")
        org_id = cur.fetchone()[0]
        
        cur.execute("""
            INSERT INTO deals (name, amount, stage, organization_id, created_at)
            VALUES (%s, %s, %s, %s, NOW())
            RETURNING id
        """, (name, amount, stage, org_id))
        deal_id = cur.fetchone()[0]
        conn.commit()
        print(f"Deal created with ID: {deal_id} for Org {org_id}")

        # 2. Get Initial AI Score (Simulate the app's initial score calculation)
        print("\n--- Step 2: Get Initial AI Score (0 tasks) ---")
        features = {
            "amount": float(amount),
            "stage": stage,
            "note_count": 0,    # No tasks yet
            "deal_age": 0,       # Initial
            "demo_completed": False,
            "champion_identified": False
        }
        res_initial = post_predict(features)
        score_initial = res_initial["ai_score"]
        print(f"Initial AI Score (0 tasks): {score_initial}%")

        cur.execute("UPDATE deals SET ai_score = %s WHERE id = %s", (score_initial, deal_id))
        conn.commit()

        # 3. Add a Task to that Deal
        print("\n--- Step 3: Add a Task to that Deal ---")
        cur.execute("""
            INSERT INTO tasks (title, deal_id, organization_id, status)
            VALUES (%s, %s, %s, %s)
        """, (f"Follow up with {name}", deal_id, org_id, "Pending"))
        conn.commit()
        print("Task added and linked to deal.")

        # 4. Re-calculate AI Score (Simulate the createTask trigger)
        print("\n--- Step 4: Re-calculate AI Score (1 task) ---")
        features["note_count"] = 1 # Now there is 1 task
        res_after = post_predict(features)
        score_after = res_after["ai_score"]
        print(f"AI Score after 1 Task: {score_after}%")

        cur.execute("UPDATE deals SET ai_score = %s WHERE id = %s", (score_after, deal_id))
        conn.commit()

        # Comparison with a bit of precision margin
        if abs(score_after - score_initial) > 0.001:
            print(f"\n✅ SUCCESS: AI Score changed from {score_initial}% to {score_after}%")
        else:
            print(f"\n❌ FAIL: AI Score remained the same ({score_initial}%). Check if the model uses note_count.")

    finally:
        cur.close()
        conn.close()

if __name__ == "__main__":
    test_flow()

File: scripts/test-task-logic.ts
import { query } from '../lib/db';

async function testLogic() {
  console.log('--- Testing Task-Deal AI Logic ---');

  try {
    // 1. Get an existing open deal
    const dealRes = await query("SELECT id, name, ai_score FROM deals WHERE stage NOT IN ('Closed Won', 'Closed Lost') LIMIT 1");
    if (dealRes.rows.length === 0) {
      console.log('No open deals found to test.');
      process.exit(0);
    }

    const deal = dealRes.rows[0];
    const initialScore = deal.ai_score;
    console.log(`Initial Deal: ${deal.name} (ID: ${deal.id}), Score: ${initialScore}`);

    // 2. Add a task linked to this deal
    console.log('Adding a linked task...');
    // We simulate the createTask logic here manually since it's a server action
    // but we want to see if the score changes.

    // Actually, I'll just run the score recalculation logic directly as it is in actions.ts
    // but triggered by a manual insert to see the effect.
    await query(
      `INSERT INTO tasks (title, description, priority, deal_id, organization_id)
       VALUES ($1, $2, $3, $4, $5)`,
      ['AI Test Task', 'Testing memory density', 'Medium', deal.id, 1]
    );

    // Now trigger the same logic as in actions.ts:createTask
    const dealDataRes = await query("SELECT * FROM deals WHERE id = $1", [deal.id]);
    const d = dealDataRes.rows[0];
    const deal_age = Math.floor((new Date().getTime() - new Date(d.created_at).getTime()) / (1000 * 60 * 60 * 24));
    const ntRes = await query("SELECT COUNT(*) as count FROM tasks WHERE deal_id = $1", [deal.id]);
    const note_count = parseInt(ntRes.rows[0].count);

    console.log(`Current Note Count for Deal ${deal.id}: ${note_count}`);

    const pythonUrl = process.env.PYTHON_API_URL || "https://crm-iogr.onrender.com";
    const pr = await fetch(`${pythonUrl}/predict`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        amount: d.amount,
        stage: d.stage,
        note_count,
        deal_age,
        demo_completed: d.demo_completed,
        champion_identified: d.champion_identified
      }),
    });

    if (pr.ok) {
      const { ai_score } = await pr.json();
      await query(`UPDATE deals SET ai_score = $1 WHERE id = $2`, [ai_score, deal.id]);

      if (ai_score !== initialScore) {
        console.log('✅ SUCCESS: AI Score updated after task addition!');
      } else {
        console.log('ℹ️ INFO: AI Score remained the same (model might need more training or notes to shift).');
      }
    } else {
      console.error('❌ FAILED: Python API error');
    }

  } catch (err) {
    console.error('Error during test:', err);
  }
  process.exit(0);
}

testLogic();
